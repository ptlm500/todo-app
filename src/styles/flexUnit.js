// based on https://nilsb.org/2019-02-12-sass-mixins-in-styled-components/
export default function flexUnit(amount, min, max, unit = 'vw', prop = 'font-size') {
  const minBreakpoint = (min / amount) * 100;
  const maxBreakpoint = max ? (max / amount) * 100 : false;
  const dimension = unit === 'vw' ? 'width' : 'height';

  return `
    @media (max-${dimension}: ${minBreakpoint}px) {
      ${prop}: ${min}px;
    }

    ${max ? `
      @media (min-${dimension}: ${maxBreakpoint}px) {
        ${prop}: ${max}px;
      }
    ` : ''}

    ${prop}: ${amount}${unit}
  `;
}
