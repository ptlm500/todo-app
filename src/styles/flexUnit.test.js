import flexUnit from './flexUnit';

const originalWarn = console.warn;
console.warn = jest.fn();

beforeEach(() => {
  jest.clearAllMocks();
});

afterAll(() => {
  console.warn = originalWarn;
});

describe('flexUnit', () => {
  it('defaults to use width and provide a font-size style', () => {
    expect(flexUnit(5, 100, 200)).toMatchSnapshot();
  });

  it('uses width when vw is provided', () => {
    expect(flexUnit(5, 100, 200, 'vw')).toMatchSnapshot();
  });

  it('uses height when vh is provided', () => {
    expect(flexUnit(5, 100, 200, 'vh')).toMatchSnapshot();
  });

  it('warns when an invalid unit is provided and defaults to width', () => {
    expect(flexUnit(5, 100, 200, 'invalid')).toMatchSnapshot();

    expect(console.warn)
      .toHaveBeenCalledWith(
        '[flexUnit]: invalid unit provided, please use \'vh\' or \'vw\'. Defaulting to width.',
        'invalid'
      );
  });

  it('allows a custom prop to be specified', () => {
    expect(flexUnit(5, 100, 200, 'vw', 'width')).toMatchSnapshot();
  });
});
