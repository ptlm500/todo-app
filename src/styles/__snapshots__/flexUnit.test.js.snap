// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flexUnit allows a custom prop to be specified 1`] = `
"
    @media (max-width: 2000px) {
      width: 100px;
    }
    
      @media (min-width: 4000px) {
        width: 200px;
      }
    
    width: 5vw
  "
`;

exports[`flexUnit defaults to use width and provide a font-size style 1`] = `
"
    @media (max-width: 2000px) {
      font-size: 100px;
    }
    
      @media (min-width: 4000px) {
        font-size: 200px;
      }
    
    font-size: 5vw
  "
`;

exports[`flexUnit uses height when vh is provided 1`] = `
"
    @media (max-height: 2000px) {
      font-size: 100px;
    }
    
      @media (min-height: 4000px) {
        font-size: 200px;
      }
    
    font-size: 5vh
  "
`;

exports[`flexUnit uses width when vw is provided 1`] = `
"
    @media (max-width: 2000px) {
      font-size: 100px;
    }
    
      @media (min-width: 4000px) {
        font-size: 200px;
      }
    
    font-size: 5vw
  "
`;

exports[`flexUnit warns when an invalid unit is provided and defaults to width 1`] = `
"
    @media (max-width: 2000px) {
      font-size: 100px;
    }
    
      @media (min-width: 4000px) {
        font-size: 200px;
      }
    
    font-size: 5vw
  "
`;
